Implement test script for  US 302-Login function

honganh1206 opened the pull request:
https://dev.azure.com/RookiesBatchVI/RookiesBatch6Group4/_workitems/edit/302

Run success (Dec 7, 2022 12:53):
![image](https://user-images.githubusercontent.com/79431746/206099619-97a46ba8-703e-4ae4-af0d-a7b4e8c0aae2.png)

**Note**: I have yet to deploy cleaning data by disabling users. I will work on that later.

honganh1206 committed:
Implement test script for Login function

daipn1995 commented on AssetManagementTestProject/PageObj/CreateUserPage.cs:
```diff
@@ -1,5 +1,5 @@
 ﻿using CoreFramework.DriverCore;
-
+using AssetManagementTestProject.DAO;
```
remove unused dependencies

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/PageObj/HomePage.cs:
```diff
@@ -18,10 +18,9 @@ public class HomePage : WebDriverAction
-    public string ReturnHomePageUrl()
+    public string ReturnHeaderHomePage()
     {
-        WaitToBeVisible(headerHomePage); // Wait for the page to log in before verifying
-        return GetUrl();
+        return headerHomePage;
     }
```
You should make the headerHomePage as a public property

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -7,27 +7,49 @@ namespace AssetManagementTestProject;
+        LoginPage.Login(Constant.STAFF_USERNAME, Constant.STAFF_PASSWORD);
+        DriverBaseAction.IsElementDisplayed(MenuBarLeft.ReturnHomeBtn());
+    }
+    [TearDown]
+    public void TearDown()
+    {
     }
```
Remove teardown if any

honganh1206 replied:
Done

honganh1206 replied:
This `TearDown` goes with the `SetUp` above, and when I remove the `SetUp` the test shows the same behavior as the comment above suggest

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -7,27 +7,49 @@ namespace AssetManagementTestProject;
     public void UserLoginFirstTime()
     {
-        LoginPage.Login(Constant.STAFF_USERNAME_1, Constant.BASE_STAFF_PASSWORD_1);
+        LoginPage.Login(Constant.STAFF_USERNAME_1ST_TIME, Constant.STAFF_PASSWORD_1ST_TIME);
```
How the STAFF_USERNAME_1ST_TIME user and STAFF_USERNAME user can exist in the default database?

daipn1995 replied:
You should create default data by sending the API and clear data after the test is done

honganh1206 replied:
I'm kinda stuck at this, so I'm going to update if I find something useful later.

daipn1995 replied:
Please contact the dev team to know how to create new users by API request

honganh1206 replied:
Done with the creating part, I will update this thread when I finish with clearing data by disabling created user

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -7,27 +7,49 @@ namespace AssetManagementTestProject;
 [TestFixture]
 public class LoginTest : NUnitWebTestSetup
 {
+    protected ChangePassword1stTimePage ChangePw1stTime;
+
+    [SetUp]
+    public void SetUp()
+    {
+        ChangePw1stTime = new ChangePassword1stTimePage();
+    }
```
Remove setup you can define the value for that variable
`protected ChangePassword1stTimePage ChangePw1stTime = new ChangePassword1stTimePage();`

honganh1206 replied:
Done

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -7,27 +7,49 @@ namespace AssetManagementTestProject;
-        LoginPage.Login(Constant.ADMIN_USERNAME_HN, Constant.BASE_ADMIN_PASSWORD);
-        Asserter.AssertStringEquals(HomePage.ReturnHomePageUrl(), Constant.BASE_URL);
+        LoginPage.Login(Constant.ADMIN_USERNAME_HN, Constant.ADMIN_PASSWORD_HN);
+        DriverBaseAction.WaitToBeVisible(HomePage.ReturnHeaderHomePage());
```
Nullable value => DriverBaseAction?

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestSetup/NUnitWebTestSetup.cs:
```diff
@@ -24,7 +23,7 @@ public void SetUp()
         Asserter = new Asserter.Asserter();
         MenuBarLeft = new LeftMenuPage();
-        ChangePw1stTime = new ChangePassword1stTimePage();
+
     }
     [TearDown]
     public void TearDown()
```
The method of a child class cannot have the same name as the parent class if you want to do some different actions
Change the name to WebBaseTearDown and WebBaseSetUp

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -7,27 +7,49 @@ namespace AssetManagementTestProject;
+        DriverBaseAction.CompareUrls(Constant.BASE_URL);
     }
     [Test]
     public void UserLoginFirstTime()
```
Test titles should be an expectation. Not an action

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -7,27 +7,49 @@ namespace AssetManagementTestProject;
-        Asserter.AssertStringEquals(HomePage.ReturnHomePageUrl(), Constant.BASE_URL);
+        LoginPage.Login(Constant.ADMIN_USERNAME_HN, Constant.ADMIN_PASSWORD_HN);
+        DriverBaseAction.WaitToBeVisible(HomePage.ReturnHeaderHomePage());
+        DriverBaseAction.CompareUrls(Constant.BASE_URL);
```
You should do an assert for the home page element UI instead of the link

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -7,27 +7,49 @@ namespace AssetManagementTestProject;
+    public void StaffCanLoginToTheApp()
+    {
+        LoginPage.Login(Constant.STAFF_USERNAME, Constant.STAFF_PASSWORD);
+        DriverBaseAction.IsElementDisplayed(MenuBarLeft.ReturnHomeBtn());
```
The expectation is a clause that defines what behavior will happen next. DriverBaseAction.IsElementDisplayed() was not a behavior
Then it should not be an expectation. So the reader cannot understand what are you trying to do.
You have to do the assert function such as: ....Element(...).IsDisplayed()
or ....ElementIsDisplayed(....)

honganh1206 replied:
I think while this is possible, but this  (....Element(...).IsDisplayed()) means using an assert function outside WebDriverAction and it will not be recorded by HtmlReport? Or did you mean to rewrite the method name from IsElementDisplayed to ElementIsDisplayed?

honganh1206 replied:
Done. I added a method in the `Asserter` class to verify whether element(s) is displayed

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -7,27 +7,49 @@ namespace AssetManagementTestProject;
-        LoginPage.Login(Constant.STAFF_USERNAME_1, Constant.BASE_STAFF_PASSWORD_1);
+        LoginPage.Login(Constant.STAFF_USERNAME_1ST_TIME, Constant.STAFF_PASSWORD_1ST_TIME);
         DriverBaseAction.IsElementDisplayed(ChangePw1stTime.AskChangePwFirstLogin());
         DriverBaseAction.CompareUrls(ChangePw1stTime.ReturnExpectedChangePw1stTimeUrl());
```
CompareUrls means that you're comparing 2 values. It does not matter even if they're different or not.
=> Do verification instead => Change the method name

honganh1206 replied:
I am not quite clear about this. Do you mean changing the method name CompareUrls to something else (which means I have to adjust the core fw) or using fluent assertion to verify?

honganh1206 replied:
Done. I created an asserter to assert two urls

daipn1995 commented on AssetManagementTestProject/PageObj/LeftMenuPage.cs:
```diff
@@ -18,13 +18,36 @@ public List<string> ReturnMenuBar()
+        {
+            menuBar.Add(btnReportInMenu);
+        }
         return menuBar;
```
You're using this method to check whether menu bar items are displayed or not.
If you want to check what the exact menu item will be displayed, please define the name.
Don't try to push `if (FindElementByXpath(btnManageUserInMenu) != null)` this field will be ignored at the step you use it to check

daipn1995 replied:
Resolved

honganh1206 committed:
Resolve most comments

honganh1206 committed:
Delete author

honganh1206 committed:
Resolve comments + Add methods and locators

Remove unnecessary usings + Rename public locators
Fix SetUp in LoginTest
Add locators + methods for Create User to test First Login/Change password
Create UserDAO object as test data to test Create User

daipn1995 commented on AssetManagementTestProject/PageObj/CreateUserPage.cs:
```diff
@@ -1,9 +1,70 @@
+    private readonly string dropBarType = "//input[contains(@id, 'formCreateUser_role')]";
+    private readonly string tfFirstName = "//input[contains(@id, 'formCreateUser_firstName')]";
+    private readonly string tfLastName = "//input[contains(@id, 'formCreateUser_lastName')]";
+    private readonly string tickGenderFemale = "(//input[contains(@type, 'radio')])[1]";
+    private readonly string tickGenderMale = "(//input[contains(@type, 'radio')])[2]";
```
Can we have an XPath cover both female and male cases based on the action text ("Female"/"Male")?

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -6,28 +6,41 @@ namespace AssetManagementTestProject;
+{    
+    protected ChangePassword1stTimePage? ChangePw1stTime; 
     [Test]
     public void UserLoginSuccess()
```
UserCanLogin

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -6,28 +6,41 @@ namespace AssetManagementTestProject;
     }
     [Test]
-    public void UserLoginFirstTime()
+    public void UserAskedChangePasswordFirstTime()
```
UserIsAskedToChangePassword...

honganh1206 replied:
Done, but somehow my vscode extension showed that you commented this three days ago, but on Github it showed an hour ago
![image](https://user-images.githubusercontent.com/79431746/206258077-e5154a8d-952f-432a-8c47-53a5471df349.png)
![image](https://user-images.githubusercontent.com/79431746/206258248-7693d5d6-e720-46ff-837f-6fdad9fb4938.png)

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -6,28 +6,41 @@ namespace AssetManagementTestProject;
-    [TestCase(Constant.ADMIN_USERNAME_HN, Constant.BASE_ADMIN_PASSWORD)]
-    public void UserCanLoginToTheApp(string userName, string password)
+    [Test] 
+    public void UserLoginWithNewPassword()
```
1. UserCanLoginWithNewPassword
2. Your test scenario should have the step to change user passwords before checking new password

honganh1206 replied:
Done. I think this comment is for the outdated code. For the latest code I create a new user with API, and then change the 1st time password before logging in again

daipn1995 commented on AssetManagementTestProject/TestCases/LoginTest.cs:
```diff
@@ -6,28 +6,41 @@ namespace AssetManagementTestProject;
-        LogoutPopup.LogOutOfPage();
+        LoginPage.Login(Constant.ADMIN_USERNAME, Constant.CHANGED_ADMIN_PASSWORD);
+        DriverBaseAction?.WaitToBeVisible(HomePage.HeaderHomePage);
+        Asserter.AssertUrlsEquals(DriverBaseAction?.GetUrl(),Constant.BASE_URL);
```
Format the code `(DriverBaseAction?.GetUrl(), Constant.BASE_URL);`

daipn1995 replied:
We should assert the header homepage is displayed by `AssertElementIsDisplayed` method.
Then, we are checking the end-user behavior, they don't view the link by default. Don't need to check that.
Follow your test steps in test cases

honganh1206 committed:
Create Authorization + ManageUser service + Log bugs

Create APITest to test calling API (Will delete before merging)
Delete private author
Add more properties to DAO
Fix Menu bar assertion

honganh1206 commented:
I am working on calling API to create default data and clear it after the test is done.

However, I ran into this bug while trying to do a GET request to get all users 
![Screenshot_20221205_101336](https://user-images.githubusercontent.com/79431746/205675636-698c1d13-ee35-4545-985d-828d92ed5891.png)

I think the problem is in this line after debugging:
https://github.com/RookiesVIFinal/group4msauto/blob/70c436c43a0f5abf9ba5bee3359198de630e9a9c/CoreFramework/APICore/APIRequest.cs#L129

I hope you can see it for me @daipn1995. Thanks a lot!

**NOTE**: I only added the APITest.cs class for testing API, I will delete it before merging.

## **UPDATE**: I made a mistake here, it should be `api/authentication` and not `api/authorization`
https://github.com/RookiesVIFinal/group4msauto/blob/70c436c43a0f5abf9ba5bee3359198de630e9a9c/AssetManagementTestProject/Services/AuthorizationService.cs#L12

However, I came across a new bug after changing the line above:

![image](https://user-images.githubusercontent.com/79431746/205698200-8b683827-7327-4e48-b838-17da8ae0784d.png)


[This stackoverflow question suggests that I should use HttpClient](https://stackoverflow.com/questions/54747260/system-net-webexception-internal-server-error-500), but I dont think it is feasible for our project.

honganh1206 committed:
Fix end point + Delete more comments

honganh1206 commented:
## UPDATE Dec 7, 2022: I can now create new users with API, but I have troubles formatting the password
![image](https://user-images.githubusercontent.com/79431746/205983783-7b3bb907-e51c-461e-8d64-0103cf09d583.png)

Here is part of the code:
![image](https://user-images.githubusercontent.com/79431746/205983904-82f0992b-a81b-4ad7-957d-602b4e106652.png)

The `string newUserDob` is "12/06/2000", so I dont get why this is not working , although I have follow [some examples](https://stackoverflow.com/questions/18831555/convert-date-string-to-another-date-string-with-different-format)
![image](https://user-images.githubusercontent.com/79431746/205984316-dc2c4aca-b8f0-4610-9886-144ecd3127ad.png)


I would love to hear some comments and advice if possible!

## Nevermind, I passed in the wrong format, it should be `"dd/MM/yy"` corresponding to the `string newUserDob`, and it should work fine now

honganh1206 committed:
Finish Login - Wait for peer review

Created DAO for different services
Renamed + Rewrote LoginTest with creating users with API
Changed WebDriver window size
Remove out-of-scope DAO and asserters

Minh-Tester commented on AssetManagementTestProject/PageObj/HomePage.cs:
```diff
@@ -6,23 +6,18 @@ public class HomePage : WebDriverAction
+    public string BtnAcceptAssignment = "(//span[contains(@aria-label, 'check')])[1]";
+    public string BtnDeclineAssignment = "(//span[contains(@aria-label, 'close')])[1]";
+    public string BtnRequestReturnAsset = "(//span[contains(@aria-label, 'undo')])[1]";
 
```
Please add #endregion

honganh1206 replied:
I added it a while ago, because without it the test cannot run

![image](https://user-images.githubusercontent.com/79431746/206141033-62b1e10a-fdfc-4a49-97c5-f58443b74eb1.png)

honganh1206 commented:
> Please add #endregion 

I already added it because without it the test cannot run.

honganh1206 committed:
Add API services + DAO

API services for get + edit + disable users
DAOs for each services

honganh1206 committed:
Add clean data + Renaming property

Added a clean data service
Renamed some properties to CamelCase
Added test data to test create user on UI

daipn1995 commented on AssetManagementTestProject/TestSetup/Constant.cs:
```diff
@@ -1,27 +1,79 @@
+    public const string ADMIN_PASSWORD_HN = "Admin@123";
+    #endregion
+    #region CREATED USERNAMES-PWS
+    public const string STAFF_USERNAME_1ST_TIME = "anhp1";
+    public const string STAFF_PASSWORD_1ST_TIME = "anhp1@29112000";
```
Remove unused variables

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestCases/US302_LoginTest.cs:
```diff
@@ -0,0 +1,84 @@
+    {
+        AuthorizationService = new AssetManagementAPIServices();
+        Token = AuthorizationService.ReturnLoginToken(Constant.ADMIN_USERNAME_HN, Constant.ADMIN_PASSWORD_HN);
+        APIService = new AssetManagementAPIServices();
```
Move this param to the parent as a protected property

honganh1206 replied:
Done. I realized that if I move Authorization + create new user with API into the WebTestSetup then after each test case, the data is cleaned

daipn1995 commented on AssetManagementTestProject/TestCases/US302_LoginTest.cs:
```diff
@@ -0,0 +1,84 @@
+    [Test]
+    public void TC02_UserAskedChangePasswordFirstTime()
+    {
+        AuthorizationService = new AssetManagementAPIServices();
+        Token = AuthorizationService.ReturnLoginToken(Constant.ADMIN_USERNAME_HN, Constant.ADMIN_PASSWORD_HN);
+        APIService = new AssetManagementAPIServices();
+        NewUser = APIService.ReturnNewUser(Constant.NEW_ADMIN_HN, Token);
+        FirstTimeLoginData newLoginData = new FirstTimeLoginData();
+        newLoginData.NewUser = NewUser;
+        string newAdminUsername = newLoginData.GetUsername();
+        string newAdminPassword = newLoginData.GetPassword();
```
This is common action => PrepareNewUser()
Can we move it to the parent test class?

honganh1206 replied:
Done

honganh1206 committed:
Resolve comments

Move Authorization + Create user with API to parent class (WebTestSetup)
Changed some properties to protected

honganh1206 committed:
Fix ViewDetailedUserDAO + Add locators

honganh1206 committed:
Renaming methods in WebAction + Adjust TakeScreenShot

honganh1206 committed:
Prepare data to create user + Add locators
